### ISSUE TITLE
Trace abstraction

### ISSUE THREAD

**Issue by shayanh** [DistributedClocks/tracing#3]
My suggestion for adding trace abstraction to the tracing library:

### `Tracer` API:
`Tracer` is the tracing client.
```
type Tracer struct {
	// No exported field
}
```

- `func NewTracerFromFile(configFile string) *Tracer`
  NewTracerFromFile instantiates a fresh tracer client from a configuration file.
- `func  NewTracer(config TracerConfig) *Tracer`
  NewTracer instantiates a fresh tracer client.
- `func (t *Tracer) StartTrace() Trace`
   StartTrace creates and starts a new trace object. Also, it records a `TraceStart` action.
- `func (t *Tracer) ReceiveToken(token TracingToken) Trace`
  ReceiveToken returns a trace by extracting the encapsulated trace ID in the token. Then records the token by calling `RecordAction` with `ReceiveTokenTrace`.
- `func (t *Tracer) Close() error`
   Close cleans up the connection to the tracing server.
- `func (t *Tracer) SetShouldPrint(shouldPrint bool)`
   SetShouldPrint determines whether its traces RecordAction should log the action being recorded as it sends the action to the tracing server.

### `Trace` API:
`Trace` is a collection of actions. Each trace has a unique ID.
For more information, please see the [OpenTracing specification](https://opentracing.io/specification/), although we don't have spans.

```
type Trace struct {
	// ID is unique identifier with 
	ID string
}
```

- `func (t *Trace) RecordAction(record interface{})`
  RecordAction ensures that the `record` is recorded by the tracing server, and optionally logs the `record`'s contents. `record` can be any struct value; its contents will be extracted via reflection. RecordAction also updates the vector clock.
- `func (t *Trace) GenerateToken() TracingToken`
   GenerateToken produces a fresh `TracingToken`, and records the event via Trace's `RecordAction`. This allows analysis of the resulting trace to correlate token generation and token reception. The generated token contains the vector clock and trace ID.
- `func (t *Trace) Finish()`
    Finish records `TraceFinish` action and finalizes Trace state. Finish() must be the last call made to any trace instance. For each trace ID, Finish must be called exactly once. In other words, you can have different trace instances with the same ID, but you should call Finish only once from one of them.
    
Note that these changes are not backward compatible. We can keep `RecordAction` and `GenerateToken`  in the Tracer, but it will be confusing and error-prone.

**Comment by bestchai** [comment]
Tracer.SetShouldPrint -- Could use more text since it's unclear if this is referring to a local log at the client or the log at the server and whether or not it controls if RecordAction is generated or if it gets logged/printed. (I think the RecordActions from Tracer should always be generated and received by the tracing server for completeness.)

Tracer.StartTrace -- I'm assuming this needs to generate a unique ID for the returned trace. I would mention this. (Are you thinking of doing this in coordination with the tracing server?)

Tracer.Close -- Does this mean that after this call, the use of any previously generated local Trace instances generates errors? i.e., would Trace.RecordAction generate an error after Trace.Close? I think that's what we'd want, but it's implicit now.

Trace.Finish -- I'm not sure about requiring this. The trace could terminate at a node because the node fails without a Finish invocation. I'm also not sure about requiring just one of these. Since we're in the vclock setting, you could imagine each process that has some portion of the trace calling Finish to indicate that it's done processing this portion of the trace. Or, we could get rid of Finish altogether; imposing ordering requirements on it is tricky.

**Comment by shayanh** [comment]
`Tracer.SetShouldPrint` -- indicates that the Tracer instance should log (print to stdout) the recorded actions or not. Tracer always sends the recorded actions to the tracing server.

`Tracer.StartTrace` -- Yes, it generates a unique ID for the returned trace. Good point, I'll add it in the description.

`Tracer.Close` -- After this call, the use of any previously generated local Trace instances leads to undefined behavior. I'm not sure it's better to return an error explicitly, since we have to add an error output to all of the Tracer methods.

`Tracer.Finish` -- I can see that requiring just once is tricky, especially in the case of failures. But I thought that it would be nice to have a corresponding `TraceFinish` event for each `TraceStart` event. Maybe we should remove both `TraceStart` and `TraceFinish` actions.

**Comment by bestchai** [comment]
Tracer.Close -- I think returning error is the more Go-accepted way of handling this. But, you're right, it'll inject error returns everywhere.

Trace.Finish (note: not Tracer.Finish)-- I think it's okay to include but not make it unique. i.e., Tracer.TraceStart is unique b/c every trace must start somewhere and are always given a unique ID, so TraceStart is easy to make unique. Trace.Finish, to my mind, could be a way to indicate that the current instance of the trace at a particular node is completed. For example, a worker could call Trace.Finish once it has registered a result or a cancel for one of the submitted pow jobs. To me, it indicates that there won't be any more events associated with this trace on this node (but there could be more events at other nodes).

I think that if we want a globally unique Finish action, then we can specify such a custom RecordAction type instead, and explain then where precisely it should be logged and its semantics (globally unique, cannot be followed by any more actions for the corresponding trace, etc).

One more thing to include is assumption about the tracing server availability. We probably have to require that it is running/available during the entire time period that tracing is being used.

**Comment by shayanh** [comment]
`Tracer.Close` -- I suggest let's go without errors to have a simpler API.

`Trace.Finish` (my bad in the previous comment): My suggestion is to remove `TraceFinish` record. Because we cannot ensure that it can indicate that there won't be any more events associated with this trace on this node. Because a node can receive a trace from another node (by using `ReceiveToken`). I think it's useless and makes things more complicated.

And yes, we assume tracing server availability. I'll include this in the documentation.

**Comment by bestchai** [comment]
Re: Because a node can receive a trace from another node (by using ReceiveToken) -- actually, that's a nice way to find bugs. If a node calls Trace.Finish in the expectation that the trace should not come back to it, and later the node receives this trace again via token, then ReceiveToken could return an error and note that the trace cannot be re-created since it was already finalized on this node.

But, I think I agree with you -- it makes the API smaller/easier and the above is hard to reason about. So, let's drop it.

**Comment by shayanh** [comment]
I agree with you about that. It's a nice way to find bugs, but it's hard to understand. So as you said, let's drop it.

Therefore I guess that the design is finalized, and I can start the implementation.
